nav2_collision_monitor/src/polygon_base.cpp

```diff
@@ -0,0 +1,175 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_collision_monitor/polygon_base.hpp"
+#include "nav2_collision_monitor/kinematics.hpp"
+
+#include "nav2_util/node_utils.hpp"
+
+namespace nav2_collision_monitor
+{
+
+PolygonBase::PolygonBase()
+: polygon_type_(POLYGON_BASE), polygon_name_(""), action_type_(DO_NOTHING),
+  stop_points_(-1), slowdown_(0.0), time_before_collision_(-1.0),
+  simulation_time_step_(0.0)
+{
+}
+
+PolygonBase::PolygonBase(
+  const nav2_util::LifecycleNode::WeakPtr & node,
+  const std::string polygon_name,
+  const double simulation_time_step)
+: node_(node), polygon_type_(POLYGON_BASE), polygon_name_(polygon_name), action_type_(DO_NOTHING),
+  stop_points_(-1), slowdown_(0.0), time_before_collision_(-1.0),
+  simulation_time_step_(simulation_time_step)
+{
+  auto node_sptr = node_.lock();
+  if (node_sptr) {
+    RCLCPP_INFO(node_sptr->get_logger(), "Creating PolygonBase");
+  }
+}
+
+PolygonBase::~PolygonBase()
+{
+  auto node = node_.lock();
+  if (node) {
+    RCLCPP_INFO(node->get_logger(), "Destroying PolygonBase");
+  }
+}
+
+bool PolygonBase::getParameters()
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  try {
+    // Get action type
+    nav2_util::declare_parameter_if_not_declared(
+      node, polygon_name_ + ".action_type", rclcpp::ParameterValue("stop"));  // Stop by default
+    const std::string at_str =
+      node->get_parameter(polygon_name_ + ".action_type").as_string();
+    if (at_str == "stop") {
+      action_type_ = STOP;
+    } else if (at_str == "slowdown") {
+      action_type_ = SLOWDOWN;
+    } else {  // at_str == "approach"
+      action_type_ = APPROACH;
+    }
+
+    if (action_type_ == STOP) {
+      nav2_util::declare_parameter_if_not_declared(
+        node, polygon_name_ + ".stop_points", rclcpp::ParameterValue(3));
+      stop_points_ = node->get_parameter(polygon_name_ + ".stop_points").as_int();
+    } else if (action_type_ == SLOWDOWN) {
+      nav2_util::declare_parameter_if_not_declared(
+        node, polygon_name_ + ".stop_points", rclcpp::ParameterValue(3));
+      stop_points_ = node->get_parameter(polygon_name_ + ".stop_points").as_int();
+      nav2_util::declare_parameter_if_not_declared(
+        node, polygon_name_ + ".slowdown", rclcpp::ParameterValue(0.5));
+      slowdown_ = node->get_parameter(polygon_name_ + ".slowdown").as_double();
+    } else {  // action_type_ == APPROACH
+      nav2_util::declare_parameter_if_not_declared(
+        node, polygon_name_ + ".time_before_collision",
+        rclcpp::ParameterValue(5.0));
+      time_before_collision_ =
+        node->get_parameter(polygon_name_ + ".time_before_collision").as_double();
+    }
+  } catch (const std::exception & ex) {
```

Comment:
In new update I've made `PolygonBase::getParameters()` routine to be virtual in order to call its appropriate implementation (`Polygon::getParameters()` or `Circle::getParameters()`) from new common for all polygons `configure()` routine.
In other words, the structure will be something like:
```
class PolygonBase
{
public:
  bool configure() {
    ...
    ret = getParameters();
  }
protected:
  virtual bool getParameters() {
    do something
  }
};

class Polygon : public PolygonBase
{
protected:
  virtual bool getParameters() {
    ret = PolygonBase::getParameters();
    do something else
  }
};

class Polygon : public PolygonBase
{
protected:
  virtual bool getParameters() {
     ret = PolygonBase::getParameters();
    do something else
  }
};
```

New `configure()` method (as well as `activate()` and `deactivate()` routines) are required to be a part of publishing each polygon work, since polygons now will contain lifecycle publishers (according to https://github.com/ros-planning/navigation2/pull/2982#discussion_r886135043). In this design,
I think making `void PolygonBase::getParameters()` and `virtual bool PolygonBase::getParameters() = 0` only to avoid if-cheks will only complicate the code.

Reply:
Done in latest update. In one place I've left return value to be `bool` but always true, to not multiply parent-child instances of `getParameters()`
https://github.com/ros-planning/navigation2/pull/2982/files#diff-63fb4f480d89fa7be1bc8ec1bf1066556ff73bad0e5a2ba97fffdcf78903e27bR132-R137
made for
https://github.com/ros-planning/navigation2/pull/2982/files#diff-d3d7f7e1c461b03d4a5868bd2766b722163399e0cb51227319a5b24e7cb8a434R61-R67

Reply:
Why not make them all void? There is no chance of throwing

Reply:
An exception could be thrown in child classes, as they have intentionally no-defaults parameters: https://github.com/ros-planning/navigation2/pull/2982#discussion_r891935150. We also could try to make something like for `PolygonBase`:
```
  void getParameters(std::string & polygon_topic);  // For parent class
  virtual bool getParameters(std::string & polygon_topic) = 0;  // For child classes
```
But in my opinion, this will only complicate the code

Reply:
If there is no polygon points set or invalid, we should be throwing an exception and exiting the process, that is an invalid condition, this node cannot operate with an invalid polygon for a zone